Add Prettier to lint command

nschonni opened the pull request:
These were removed from the Huskier PR, so resubmitting to review separately

nschonni commented:
After this or with this one, we could add the prettier call to the Auto PR fix job

caugner reviewed (approved):
LGTM, but the only problem I see is that those "already fixed" files don't get auto-fixed via lint-staged yet, so the failing test may be a surprise for contributors.

teoli2003 commented:
We want to discuss the plan to move forward with Prettier once everybody returns from the all-hand Mozilla trip (for some), and holidays (for others).

Meanwhile, we can fix content to match the output of Prettier, getting experience with the limitations of the tool (weirdness with inline elements, CDATA and MathML code "minimization", a few oddities with CSS, …), but we shouldn't move forward with automation before the plan has been discussed on mdn/mdn-community/discussions and agreed.

nschonni commented:
> LGTM, but the only problem I see is that those "already fixed" files don't get auto-fixed via lint-staged yet, so the failing test may be a surprise for contributors.

When these folders are opted back in, lint-stage/husky will autofix any regressions going forward

nschonni commented:
Gave this a rebase, and updated the globs for the global job and scripts to run for prettier changes. Still green with the now opted in folders

nschonni commented:
@queengooborg @bsmth noticed that #24046 landed. Maybe you're ready to look at this now

bsmth commented on .prettierignore:
```diff
@@ -6,7 +6,15 @@ build/
-!files/en-us/mdn/**/*.md
-!files/en-us/web/http/**/*.md
+!/*.md
+!/files/en-us/games/**/*.md
```
What do you think about getting rid of the leading slash for these paths? e.g.

```
!files/en-us/games/**/*.md
```

nschonni replied:
I kept if consistent because the `/` is needed for the root path, and the absolute gives an explicit match

bsmth replied:
OK good to know. If this is the preferred way, we should continue with that convention when adding other paths (note to self & @queengooborg).

bsmth commented on .prettierignore:
```diff
@@ -6,7 +6,15 @@ build/
+!/files/en-us/mdn/**/*.md
+!/files/en-us/related/**/*.md
+!/files/en-us/web/http/**/*.md
+!/files/en-us/webassembly/**/*.md
```
Q: Have `/related`, `/web`, `/games` and `/webassembly` had a full pass already?

nschonni replied:
yeah, I did them awhile ago before this PR. It's also checked by the green CI here

bsmth commented on .github/workflows/pr-check_markdownlint.yml:
```diff
@@ -35,11 +35,36 @@ jobs:
+
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v29.0.5
```
I think we had consistent problems with this action that weren't resolved after fixes landed in that repo. We [moved to this approach](https://github.com/mdn/content/blob/221a88897598f0ad73f3e80c4a6211992264dcf5/.github/workflows/pr-check_markdownlint.yml#L17-L29) although I dislike copy pasting it everywhere. Added some [initial thoughts on it here](https://github.com/mdn/translated-content/pull/4191#discussion_r1095532187).

nschonni replied:
Right, this was from before that switch over, I'll take a look at the other one and use the new matcher

nschonni committed:
chore: Opt-in cleaned folders and CI for Prettier

nschonni commented:
@teoli2003 let me know when to rebase this, as I'm picking up conflicts with all the PRs your landing right now

nschonni commented:
@queengooborg @teoli2003 @bsmth note the failure from a previously cleaned up folder. We need this CI backing or things will continue to creep back in.

queengooborg commented on .github/workflows/pr-check_markdownlint.yml:
```diff
@@ -35,11 +35,44 @@ jobs:
 
       - name: Lint markdown files
         run: |
           echo "::add-matcher::.github/workflows/markdownlint-problem-matcher.json"
           files_to_lint="${{ env.DIFF_DOCUMENTS }}"
           yarn markdownlint-cli2 ${files_to_lint}
+
+  prettier:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Get changed files
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          BASE_SHA: ${{ github.event.pull_request.base.sha }}
+          HEAD_SHA: ${{ github.event.pull_request.head.sha }}
+        run: |
+          # Use the GitHub API to get the list of changed files
+          # documenation: https://docs.github.com/rest/commits/commits#compare-two-commits
+          DIFF_DOCUMENTS=$(gh api repos/{owner}/{repo}/compare/${{ env.BASE_SHA }}...${{ env.HEAD_SHA }} \
+            --jq '.files | .[] | select(.status|IN("added", "modified", "renamed", "copied", "changed")) | .filename')
+          # filter out files that are not markdown
+          DIFF_DOCUMENTS=$(echo "${DIFF_DOCUMENTS}" | egrep -i "^files/.*\.md$" | xargs)
+          echo "DIFF_DOCUMENTS=${DIFF_DOCUMENTS}" >> $GITHUB_ENV
+
+      - name: Setup Node.js environment
+        uses: actions/setup-node@v3
+        with:
+          node-version: "16"
+          cache: yarn
+
+      - name: Install all yarn packages
+        run: yarn --frozen-lockfile
+
+      - name: Prettier markdown files
+        run: |
+          files_to_lint="${{ steps.changed-files.outputs.all_changed_files }}"
+          yarn prettier -c ${files_to_lint}
```
```suggestion
      - name: Lint markdown files
        run: |
          echo "::add-matcher::.github/workflows/markdownlint-problem-matcher.json"
          yarn markdownlint-cli2 ${{ steps.changed-files.outputs.all_changed_files }}

      - name: Prettier markdown files
        run: yarn prettier -c ${{ steps.changed-files.outputs.all_changed_files }}
```

queengooborg reviewed (changes requested):
There are some additional changes included in this PR that aren't entirely about just opting into using Prettier to scan the files, such as adding the slashes into the `.prettierignore`, sorting entries in other workflows and renaming jobs.

Additionally, I think that we can combine the Markdownlint and Prettier command runs into a single workflow job.

nschonni commented:
> There are some additional changes included in this PR that aren't entirely about just opting into using Prettier to scan the files, such as adding the slashes into the `.prettierignore`, sorting entries in other workflows and renaming jobs.

I've minimized the diff on the other job to just add the trigger on `.prettier*` changes. For the slash discussion, i've "unresolved the previous discussion about that one so it's visible again


> Additionally, I think that we can combine the Markdownlint and Prettier command runs into a single workflow job.

I added your suggestion, then I realized there was a bug from when I had changed from the old "diff files" action to the current one, so I rebased, but kept the collasped single job

queengooborg commented:
> I've minimized the diff on the other job to just add the trigger on .prettier* changes.

Thank you, much appreciated!

> For the slash discussion, i've "unresolved the previous discussion about that one so it's visible again

Sounds good!  I opened up #24338 to add them into our config; just waiting on review for it!

> added your suggestion, then I realized there was a bug from when I had changed from the old "diff files" action to the current one, so I rebased, but kept the collasped single job

👍

nschonni commented:
@queengooborg I cherry-picked your ignore changes from https://github.com/mdn/content/pull/24338 and fixed an issue on one of the ignores. I fixed the single unrelated failing file, but it would get fixed by the auto-PR job after this anyway

queengooborg committed:
Merge branch 'main' into prettier-ci

queengooborg committed:
Format remaining files

queengooborg committed:
Format remaining file

queengooborg reviewed (approved):
Since there was only six files left in need of formatting, I've gone ahead and just formatted them now so we can get this landed!
